examples/credential-registry/src/lib.rs

```diff
@@ -0,0 +1,1292 @@
+//! This smart contract implements an example on-chain registry for the public
+//! part of verifiable credentials (VCs).
+//!
+//! # Description
+//!
+//! The contract keeps track of credential public data, allows for managing the
+//! VC life cycle and querying VCs data and status. The intended users are
+//! issuers of VCs, holders of VCs, revocation authorities and verifiers.
+//!
+//! ## Issuer's  functionality
+//!
+//! - registration of a new credential;
+//! - updating an existing credential;
+//! - revoking a credential;
+//! - register a public key;
+//! - register a revocation authority key.
+//!
+//! Revocation authorities are some entities chosen by the issuer that has
+//! revocation capabilities. Their public keys are registered by the issuer and
+//! a revocation authority signs a revocation message with the corresponding
+//! secret key.
+//!
+//! ## Holder's functionality
+//!
+//! - revoking a credential by signing a revocation message.
+//!
+//! Revocation authority's functionality
+//!
+//! - revoking a credential by signing a revocation message.
+//!
+//! ## Verifier's functionality
+//!
+//! - view credential status to verify VC validity.
+//! - view credential data to verify proofs (verifiable presentations) requested
+//!   from holders.
+use concordium_cis2::MetadataUrl;
+use concordium_std::*;
+use core::fmt::Debug;
+
+#[derive(Serialize, SchemaType)]
+struct RegisterCredentialEvent {
+    credential_id: Uuidv4,
+    holder_id:     PublicKeyEd25519,
+    schema_ref:    String,
+}
+
+#[derive(Serialize, SchemaType)]
+struct UpdateCredentialEvent {
+    credential_id: Uuidv4,
+    holder_id:     PublicKeyEd25519,
+    schema_ref:    String,
+}
+
+#[derive(Serialize, SchemaType)]
+enum Revoker {
+    Issuer,
+    Holder,
+    Other(PublicKeyEd25519),
+}
+
+#[derive(Serialize, SchemaType)]
+struct RevokeCredentialEvent {
+    credential_id: Uuidv4,
+    holder_id:     PublicKeyEd25519,
+    revoker:       Revoker,
+    reason:        Option<String>,
+}
+
+/// The type for a credential identifier.
+/// The uuidv4 identifier is generated externally by the issuer.
+type Uuidv4 = u128;
+
+#[derive(Serialize, SchemaType, PartialEq, Eq, Clone, Copy, Debug)]
+pub enum CredentialStatus {
+    Active,
+    Revoked,
+    Expired,
+    NotActivated,
+}
+
+/// Public verifiable credential data
+#[derive(Serialize, SchemaType, PartialEq, Eq, Clone, Debug)]
+pub struct CredentialData {
+    holder_id:            PublicKeyEd25519,
+    is_holder_revocable:  bool,
+    commitment:           Vec<u8>,
```

Comment:
Do the commitments have a fixed length? If yes, we could define a type that has exactly that length.